crates/iceberg/src/expr/visitors/manifest_evaluator.rs

```diff
@@ -98,103 +101,249 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
         Ok(self.field_summary_for_reference(reference).contains_null)
     }
 
+    #[allow(clippy::bool_comparison)]
     fn not_null(
         &mut self,
         reference: &BoundReference,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+
+        // contains_null encodes whether at least one partition value is null,
+        // lowerBound is null if all partition values are null
+        let mut all_null: bool = field.contains_null && field.lower_bound.is_none();
+
+        if all_null && reference.field().field_type.is_floating_type() {
+            // floating point types may include NaN values, which we check separately.
+            // In case bounds don't include NaN value, contains_nan needs to be checked against.
+            all_null = field.contains_nan.is_some_and(|x| x == false);
+        }
+
+        Ok(!all_null)
     }
 
     fn is_nan(
         &mut self,
         reference: &BoundReference,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        Ok(self
-            .field_summary_for_reference(reference)
-            .contains_nan
-            .is_some())
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+        if let Some(contains_nan) = field.contains_nan {
+            Ok(contains_nan)
+        } else {
+            ROWS_MIGHT_MATCH
```

Comment:
The Java implementation also tests if all values are null at this point - see here: https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java#L151

You could extract [some of the code in not_null](https://github.com/apache/iceberg-rust/pull/367/files?diff=unified&w=0#diff-f3185efcdea2022bdedc017f7859727f2e44ac0c004f08ad6a96fdfa5fa6f69aR112-R120) into a shared method, `contains_nulls_only`, and call that here.